lib/web/fetch/index.js

```diff
@@ -120,8 +121,53 @@ class Fetch extends EE {
   }
 }
 
+// hoisting subscribers variables
+const subscribers = {
+  start: false,
+  end: false,
+  asyncStart: false,
+  asyncEnd: false,
+  fetchError: false
+}
+
+// publishIfSubscribed will check if there are subscribers to the channel
+// add set additional context before publishing the event
+function publishIfSubscribed (name, context, error) {
+  if (subscribers[name]) {
+    if (error) context.error = error
+    channels[name].publish(context)
+    if (error) context.error = null
+  }
+}
+
+// handlePromise will handle both fulfilled/rejected promises from p
+// for diagnostics channel it will then check for subscribers and
+// publish events accordingly
+function handlePromise (p, context) {
+  return p.promise.then(
+    (success) => {
+      publishIfSubscribed('asyncStart', context)
+      publishIfSubscribed('asyncEnd', context)
+      return success
+    },
+    (failure) => {
+      publishIfSubscribed('asyncStart', context, failure)
+      publishIfSubscribed('asyncEnd', context, failure)
+      return Promise.reject(failure)
+    })
+}
+
 // https://fetch.spec.whatwg.org/#fetch-method
-function fetch (input, init = undefined) {
+function fetch (input, init = {}) {
+  const context = { input, init, result: null, error: null }
+
+  // checking for subscribers for diagnostics channel events
+  subscribers.start = channels.start.hasSubscribers
+  subscribers.end = channels.end.hasSubscribers
+  subscribers.asyncStart = channels.asyncStart.hasSubscribers
+  subscribers.asyncEnd = channels.asyncEnd.hasSubscribers
+  subscribers.fetchError = channels.fetchError.hasSubscribers
```

Comment:
Why is subscriber state pre-cached on each call? It could change state between here and when the publish attempt would actually happen. Would it not be better to just do `if (channels[name].hasSubscribers)` rather than `if (subscribers[name])` in `publishIfSubscribed`?

Reply:
going to update this to not check for subscribers early in the case that the subscribers are not yet subscribed.